building-and-deploying-a-fastapi-app.md

```diff
@@ -0,0 +1,541 @@
+
+The `RecipeSchema` has a subclass `Config`. The subclass contains an object variable, `schema_extra`, which includes a key `example` used as the mock data in the interactive documentation.
+
+The `UpdateRecipeSchema`, on the other hand, is the model for guiding request body sent through the UPDATE route. The difference between this model and the `RecipeSchema` is that the schema values can be passed optionally.
+
+With the model in place, let's write the code for the application's routes. In `app/api.py`, add the recipes database before the base route __"/"__:
+
+```py
+recipes = [
+    {
+        "id": 1,
+        "name": "Donuts",
+        "ingredients": ["Flour", "Milk", "Sugar", "Vegetable Oil"]
+    }
+]
+```
+
+Below the base route, add the GET routes:
+
+```
+@app.get("/recipe", tags=["Recipe"])
+def get_recipes() -> dict:
+    return {
+        "data": recipes
+    }
+
+@app.get("/recipe/{id}", tags=["Recipe"])
+def get_recipe(id: int) -> dict:
+    if id > len(recipes) or id < 1:
+        return {
+            "error": "Invalid ID passed."
+        }
+
+    for recipe in recipes:
+        if recipe['id'] == id:
+            return {
+                "data": [
+                    recipe
+                ]
+            }
+
+    return {
+        "error": "No such recipe with ID {} exist".format(id)
+    }
+```
+
+You defined GET routes for retrieving all the recipes and a single recipe using its ID in the code block above. An error message is returned if an invalid ID is passed as a parameter in the __/recipe/{id}__ route. Test the routes by visiting http://localhost:8080/recipe and http://localhost:8080/recipe/1:
+
+![Recipes](https://res.cloudinary.com/adeshina/image/upload/v1608635876/zree0awqm8co1g2kf339.png)
+    
+
+![First Recipe](https://res.cloudinary.com/adeshina/image/upload/v1608635909/gdted1ywrnnkjupy9x2o.png)
+
+Define the POST route for adding new recipes:
+
+Start by updating the imports:
+
+```py
+from fastapi import FastAPI, Body
+
+from app.model import RecipeSchema, UpdateRecipeSchema
+from fastapi.encoders import jsonable_encoder
+```
+
+Next, add the POST route beneath the GET routes:
+
+```py
+@app.post("/recipe", tags=["Recipe"])
+def add_recipe(recipe: RecipeSchema = Body(...)) -> dict:
+    recipe.id = len(recipes) + 1
+    recipes.append(recipe.dict())
+    return {
+        "message": "Recipe added successfully."
+    }
+```
+
+In the code block above, we made sure the request body is modeled in our `RecipeSchema` by setting the type `recipe` to `RecipeSchema` in the `add_recipe` function. The `Body(...)` imported from FastAPI ensures that the request body is passed.
```

Comment:
Could also quickly mention what happens if the content of the request doesn't match the type?

Reply:
Added.

Reply:
thanks!